python-backend/graph_utils.py

```diff
@@ -13,74 +13,180 @@
+        if (n >= 1
+                and values[n] is not None
+                and q_values[n] is not None
+                and q_values[n] > 0
+                and q_values[n] != 1
+                and str(q_values[n]) != '0.0' and str(values[n] - limit) != '0.0'):
+
+            if mpmath.almosteq(limit, values[n]):
+                break
+
+            error = mpmath.log10(mpmath.fabs(mpmath.fsub(values[n], limit)))
+            q = mpmath.log10(mpmath.fabs(q_values[n]))
+
+            if mpmath.almosteq(q, mpmath.mpf(0)):
+                break
+
+            y_value = mpmath.fsub(-1, (mpmath.fdiv(error, q)))
 
-        x_y_pairs.append(Point2D(x=i, y=str(y_value)))
+            if not mpmath.isinf(y_value):
+                if n <= DEBUG_LINES:
+                    logger.debug(
+                        f"Delta at {n}: - 1 - log10(|{values[n]} - {limit}|) / log10(|{q_values[n]}|) = {y_value}")
+                x_y_pairs.append(Point2D(x=n, y=str(y_value)))
 
     return x_y_pairs
 
 
-def delta_coordinates(values: list[mpmath.mpf], q_values: list[mpmath.mpf], limit: mpmath.mpf) -> list[Point2D]:
+def reduced_delta_coordinates(values: OrderedDict[int, mpmath.mpf],
+                              p_values: OrderedDict[int, mpmath.mpf],
+                              q_values: OrderedDict[int, mpmath.mpf],
+                              limit: mpmath.mpf) -> list[Point2D]:
     """
     graph coords for the error delta for an expression -1 * (log(|Pn/Qn - L|) / log(Qn)) - 1
     :param values: computed values for continued fraction
     :param limit: the limit of the expression as computed by PCF(a_n, b_n).limit()
+    :param p_values: computed values for continued fraction - numerator only
     :param q_values: computed values for continued fraction - denominator only
     :return: array of [x,y] pairs for graphing purposes
     """
     x_y_pairs = []
     # graph coords of error delta: -1 * (log(|Pn/Qn - L|) / log(Qn)) - 1
-    for i in range(1, min(len(values), len(q_values))):
+    for n in values.keys():
         # we test for values that would generate irrational results or exceptions
         # e.g. divide by zero when taking log10 of 1
         # taking log 10 gives us the order of magnitude of the difference -  the number of zeros after the decimal
         # which is a gauge of the proximity of the value at n to the "limit"
         # in this case we are then comparing that precision to the precision of the denominator
-        if values[i] is not None and q_values[i] is not None and q_values[i] > 0 and q_values[i] != 1:
-            y_value = (mpmath.mpf(-1) *
-                       (mpmath.log10(abs(values[i] - limit)) / mpmath.log10(q_values[i]))
-                       - mpmath.mpf(1))
-            if i <= DEBUG_LINES:
-                logger.debug(f"Delta {i} {y_value}")
-
-            x_y_pairs.append(Point2D(x=i, y=str(y_value)))
+        if (n >= 1
```

Comment:
The processing conditions in this code looks very similar to the processing conditions in the code above. Is there an opportunity to invoke a common generator function that yields the values that meet these conditions?

Reply:
I will tackle this in a subsequent PR because I want to consolidate these functions.